### PR TITLE
Provide diagnostic suggestion in ExprUseVisitor Delegate

### LINKED ISSUE #20: Provide diagnostic suggestion in ExprUseVisitor Delegate
The [Delegate trait](https://github.com/rust-lang/rust/blob/981346fc07dd5ef414c5b1b21999f7604cece006/compiler/rustc_typeck/src/expr_use_visitor.rs#L28-L38) currently use `PlaceWithHirId` which is composed of Hir `Place` and the corresponding expression id. 

Even though this is an accurate way of expressing how a Place is used, it can cause confusion during diagnostics.

Eg:

```
let arr : [String; 5];

let [a, ...]     =   arr;
 ^^^ E1 ^^^      =  ^^E2^^
```

Here `arr` is moved because of the binding created E1. However, when we point to E1 in diagnostics with the message `arr` was moved, it can be confusing.  Rather we would like to report E2 to the user.

We want to modify the Delegate methods to provide an expression that should be used for diagnostics. 

We have tried to previously achieve something similar before https://github.com/rust-lang/rust/pull/75933/files#diff-1f18144176a6e44f905262991bed987098348cdbd688ac76f8c915ec91b7fd58R316-R319

### PR DESCRIPTION
**Pull request by null-sleep**
The [Delegate trait](https://github.com/rust-lang/rust/blob/981346fc07dd5ef414c5b1b21999f7604cece006/compiler/rustc_typeck/src/expr_use_visitor.rs#L28-L38) currently use `PlaceWithHirId` which is composed of Hir `Place` and the
corresponding expression id.

Even though this is an accurate way of expressing how a Place is used,
it can cause confusion during diagnostics.

Eg:

```
let arr : [String; 5];

let [a, ...]     =   arr;
 ^^^ E1 ^^^      =  ^^E2^^
 ```

 Here `arr` is moved because of the binding created E1. However, when we
 point to E1 in diagnostics with the message `arr` was moved, it can be
 confusing.  Rather we would like to report E2 to the user.

 Closes: https://github.com/rust-lang/project-rfc-2229/issues/20

r? @ghost

### CODE REVIEW

**Comment by null-sleep** [comment]
Making sure clippy tests pass :)

File: compiler/rustc_typeck/src/expr_use_visitor.rs [hunk 1]
```diff
@@ -27,14 +27,38 @@ use rustc_span::Span;
 /// employing the ExprUseVisitor.
 pub trait Delegate<'tcx> {
     // The value found at `place` is either copied or moved, depending
-    // on mode.
-    fn consume(&mut self, place_with_id: &PlaceWithHirId<'tcx>, mode: ConsumeMode);
+    // on mode. Where `diag_expr_id` is the id used for diagnostics for `place`.
+    //
+    // The reson for a separate expr_id for diagonostics is to support cases
```
Comment (nikomatsakis) [inline compiler/rustc_typeck/src/expr_use_visitor.rs:32]: s/reson/reason/ (I can't seem to find the "automated suggestion" UI in github anymore...)

File: compiler/rustc_typeck/src/expr_use_visitor.rs [hunk 2]
```diff
@@ -27,14 +27,38 @@ use rustc_span::Span;
 /// employing the ExprUseVisitor.
 pub trait Delegate<'tcx> {
     // The value found at `place` is either copied or moved, depending
-    // on mode.
-    fn consume(&mut self, place_with_id: &PlaceWithHirId<'tcx>, mode: ConsumeMode);
+    // on mode. Where `diag_expr_id` is the id used for diagnostics for `place`.
```
Comment (nikomatsakis) [inline compiler/rustc_typeck/src/expr_use_visitor.rs:30]: This doesn't seem grammatically correct as a sentence. Perhaps:

The value found at `place` is either copied or moved, depending on `mode`. 

The parameter `diag_expr_id` indicates the HIR id that ought to be used for diagnostics. Around pattern matching such as `let pat = expr`, the diagnostic id will be the id of the expression `expr` but the place itself will have the id of the binding in the pattern `pat`.

File: compiler/rustc_typeck/src/expr_use_visitor.rs [hunk 3]
```diff
@@ -27,14 +27,38 @@ use rustc_span::Span;
...
+    );
 
     // The value found at `place` is being borrowed with kind `bk`.
-    fn borrow(&mut self, place_with_id: &PlaceWithHirId<'tcx>, bk: ty::BorrowKind);
-
-    // The path at `place_with_id` is being assigned to.
-    fn mutate(&mut self, assignee_place: &PlaceWithHirId<'tcx>);
+    // `diag_expr_id` is the id used for diagnostics for `place`.
+    //
+    // The reson for a separate expr_id for diagonostics is to support cases
```
Comment (nikomatsakis) [inline compiler/rustc_typeck/src/expr_use_visitor.rs:45]: I'd probably write something like:

`diag_expr_id` is the id used for diagnostics (see `consume` for more details).

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Looks good, I left a few wording nits.

File: compiler/rustc_typeck/src/expr_use_visitor.rs (hunk 1 above)
Comment (arora-aman) [reply compiler/rustc_typeck/src/expr_use_visitor.rs]: image

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**